File: swift-bridge/Sources/StoreKitBridge/PurchaseOption.swift
import Foundation
import StoreKit

struct SKPurchaseOptionPayload: Codable {
    let kind: String
    let appAccountToken: String?
    let quantity: Int?
    let simulateAskToBuyInSandbox: Bool?
    let key: String?
    let value: String?
    let doubleValue: Double?
    let boolValue: Bool?
    let valueBase64: String?
    let offerId: String?
    let keyId: String?
    let nonce: String?
    let signatureBase64: String?
    let timestamp: Int?
    let compactJws: String?
    let shouldContinuePurchase: Bool?
}

func skBuildPurchaseOptions(
    from payloads: [SKPurchaseOptionPayload],
    product: Product
) throws -> Set<Product.PurchaseOption> {
    var options = Set<Product.PurchaseOption>()
    for payload in payloads {
        switch payload.kind {
        case "appAccountToken":
            guard let token = payload.appAccountToken,
                  let uuid = UUID(uuidString: token)
            else {
                throw SKBridgeError.invalidArgument("purchase option appAccountToken requires a valid UUID")
            }
            options.insert(.appAccountToken(uuid))
        case "quantity":
            guard let quantity = payload.quantity else {
                throw SKBridgeError.invalidArgument("purchase option quantity requires an integer value")
            }
            options.insert(.quantity(quantity))
        case "simulatesAskToBuyInSandbox":
            options.insert(.simulatesAskToBuyInSandbox(payload.simulateAskToBuyInSandbox ?? false))
        case "customString":
            guard let key = payload.key, let value = payload.value else {
                throw SKBridgeError.invalidArgument("customString requires key and value")
            }
            options.insert(.custom(key: key, value: value))
        case "customNumber":
            guard let key = payload.key, let value = payload.doubleValue else {
                throw SKBridgeError.invalidArgument("customNumber requires key and doubleValue")
            }
            options.insert(.custom(key: key, value: value))
        case "customBool":
            guard let key = payload.key, let value = payload.boolValue else {
                throw SKBridgeError.invalidArgument("customBool requires key and boolValue")
            }
            options.insert(.custom(key: key, value: value))
        case "customData":
            guard let key = payload.key,
                  let encoded = payload.valueBase64,
                  let data = Data(base64Encoded: encoded)
            else {
                throw SKBridgeError.invalidArgument("customData requires key and valid base64 data")
            }
            options.insert(.custom(key: key, value: data))
        case "promotionalOfferSignature":
            guard let offerID = payload.offerId,
                  let keyID = payload.keyId,
                  let nonceString = payload.nonce,
                  let nonce = UUID(uuidString: nonceString),
                  let signatureBase64 = payload.signatureBase64,
                  let signature = Data(base64Encoded: signatureBase64),
                  let timestamp = payload.timestamp
            else {
                throw SKBridgeError.invalidArgument(
                    "promotionalOfferSignature requires offerID, keyID, nonce, signatureBase64, and timestamp"
                )
            }
            options.insert(
                .promotionalOffer(
                    offerID: offerID,
                    keyID: keyID,
                    nonce: nonce,
                    signature: signature,
                    timestamp: timestamp
                )
            )
        case "promotionalOfferCompactJws":
            guard let offerID = payload.offerId,
                  let compactJWS = payload.compactJws
            else {
                throw SKBridgeError.invalidArgument(
                    "promotionalOfferCompactJws requires offerID and compactJWS"
                )
            }
            for option in Product.PurchaseOption.promotionalOffer(offerID, compactJWS: compactJWS) {
                options.insert(option)
            }
        case "introductoryOfferEligibility":
            guard let compactJWS = payload.compactJws else {
                throw SKBridgeError.invalidArgument(
                    "introductoryOfferEligibility requires compactJWS"
                )
            }
            options.insert(.introductoryOfferEligibility(compactJWS: compactJWS))
        case "winBackOffer":
            guard #available(macOS 15.0, *) else {
                throw SKBridgeError.notSupported("win-back offers require macOS 15.0+")
            }
            guard let offerID = payload.offerId else {
                throw SKBridgeError.invalidArgument("winBackOffer requires offerID")
            }
            guard let offer = product.subscription?.winBackOffers.first(where: { $0.id == offerID }) else {
                throw SKBridgeError.invalidArgument("win-back offer '\(offerID)' was not found on the product")
            }
            options.insert(.winBackOffer(offer))
        case "onStorefrontChange":
            options.insert(.onStorefrontChange { _ in payload.shouldContinuePurchase ?? true })
        default:
            throw SKBridgeError.invalidArgument("unsupported purchase option kind '\(payload.kind)'")
        }
    }
    return options
}

File: src/purchase_option.rs
use serde::{Deserialize, Serialize};

use crate::error::StoreKitError;
use crate::transaction::{Transaction, TransactionPayload};
use crate::verification_result::{VerificationResult, VerificationResultPayload};

#[derive(Debug, Clone, PartialEq, Serialize)]
#[serde(tag = "kind", rename_all = "camelCase")]
/// Represents options passed to `StoreKit.Product.purchase(options:)`.
pub enum PurchaseOption {
    /// Represents the `AppAccountToken` `StoreKit` case.
    AppAccountToken {
        /// App account token reported by `StoreKit`.
        app_account_token: String,
    },
    /// Represents the `Quantity` `StoreKit` case.
    Quantity {
        /// Value forwarded to `StoreKit` for `quantity`.
        quantity: i64,
    },
    /// Represents the `SimulatesAskToBuyInSandbox` `StoreKit` case.
    SimulatesAskToBuyInSandbox {
        /// Value forwarded to `StoreKit` for `simulate_ask_to_buy_in_sandbox`.
        simulate_ask_to_buy_in_sandbox: bool,
    },
    /// Represents the `CustomString` `StoreKit` case.
    CustomString {
        /// Value forwarded to `StoreKit` for `key`.
        key: String,
        /// Value returned by `StoreKit`.
        value: String,
    },
    /// Represents the `CustomNumber` `StoreKit` case.
    CustomNumber {
        /// Value forwarded to `StoreKit` for `key`.
        key: String,
        /// Value returned by `StoreKit`.
        value: f64,
    },
    /// Represents the `CustomBool` `StoreKit` case.
    CustomBool {
        /// Value forwarded to `StoreKit` for `key`.
        key: String,
        /// Value returned by `StoreKit`.
        value: bool,
    },
    /// Represents the `CustomData` `StoreKit` case.
    CustomData {
        /// Value forwarded to `StoreKit` for `key`.
        key: String,
        /// Value forwarded to `StoreKit` for `value_base64`.
        value_base64: String,
    },
    /// Represents the `PromotionalOfferSignature` `StoreKit` case.
    PromotionalOfferSignature {
        /// Value forwarded to `StoreKit` for `offer_id`.
        offer_id: String,
        /// Value forwarded to `StoreKit` for `key_id`.
        key_id: String,
        /// Value forwarded to `StoreKit` for `nonce`.
        nonce: String,
        /// Value forwarded to `StoreKit` for `signature_base64`.
        signature_base64: String,
        /// Value forwarded to `StoreKit` for `timestamp`.
        timestamp: i64,
    },
    /// Represents the `PromotionalOfferCompactJws` `StoreKit` case.
    PromotionalOfferCompactJws {
        /// Value forwarded to `StoreKit` for `offer_id`.
        offer_id: String,
        /// Value forwarded to `StoreKit` for `compact_jws`.
        compact_jws: String,
    },
    /// Represents the `IntroductoryOfferEligibility` `StoreKit` case.
    IntroductoryOfferEligibility {
        /// Value forwarded to `StoreKit` for `compact_jws`.
        compact_jws: String,
    },
    /// Represents the `WinBackOffer` `StoreKit` case.
    WinBackOffer {
        /// Value forwarded to `StoreKit` for `offer_id`.
        offer_id: String,
    },
    /// Represents the `OnStorefrontChange` `StoreKit` case.
    OnStorefrontChange {
        /// Value forwarded to `StoreKit` for `should_continue_purchase`.
        should_continue_purchase: bool,
    },
}

#[allow(clippy::large_enum_variant)]
#[derive(Debug)]
/// Represents the result returned by `StoreKit.Product.purchase(options:)`.
pub enum PurchaseResult {
    /// The `StoreKit` operation succeeded.
    Success(VerificationResult<Transaction>),
    /// The person cancelled the `StoreKit` flow.
    UserCancelled,
    /// The `StoreKit` flow is pending further action.
    Pending,
}

#[allow(clippy::unsafe_derive_deserialize)]
#[derive(Debug, Deserialize)]
pub(crate) struct PurchaseResultPayload {
    kind: String,
    #[serde(rename = "verificationResult")]
    verification_result: Option<VerificationResultPayload<TransactionPayload>>,
}

impl PurchaseResultPayload {
    pub(crate) fn into_purchase_result(
        self,
        transaction_handle: *mut core::ffi::c_void,
    ) -> Result<PurchaseResult, StoreKitError> {
        match self.kind.as_str() {
            "success" => {
                let verification_result = self.verification_result.ok_or_else(|| {
                    StoreKitError::Unknown(
                        "StoreKit reported a successful purchase without a verification result"
                            .to_owned(),
                    )
                })?;
                let transaction = verification_result.into_result(|payload| {
                    Transaction::from_raw_parts(transaction_handle, payload)
                })?;
                Ok(PurchaseResult::Success(transaction))
            }
            "userCancelled" => {
                if !transaction_handle.is_null() {
                    unsafe { crate::ffi::sk_transaction_release(transaction_handle) };
                }
                Ok(PurchaseResult::UserCancelled)
            }
            "pending" => {
                if !transaction_handle.is_null() {
                    unsafe { crate::ffi::sk_transaction_release(transaction_handle) };
                }
                Ok(PurchaseResult::Pending)
            }
            other => {
                if !transaction_handle.is_null() {
                    unsafe { crate::ffi::sk_transaction_release(transaction_handle) };
                }
                Err(StoreKitError::Unknown(format!(
                    "StoreKit returned an unknown purchase result kind '{other}'"
                )))
            }
        }
    }
}
